### PR TITLE
Reduce metadata IO during segment allocation

### PR DESCRIPTION
**Pull request by kfaraz**
Revives #17420 

### Description

The changes here try to reduce metadata IO during segment allocation by fetching segment payloads
for only the required segments. For other segments, only fetching the segment ID suffices.

### Changes

- Add Overlord runtime property `druid.indexer.tasklock.batchAllocationReduceMetadataIO`.
- Setting this flag to true (default value) allows the Overlord to fetch only necessary segment payloads during segment allocation
- Settings this flag to false restores original segment allocation behaviour

### Release note

Add Overlord runtime property `druid.indexer.tasklock.batchAllocationReduceMetadataIO`.
Setting this flag to true (default value) allows the Overlord to fetch only necessary segment payloads during segment allocation.

---

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by kfaraz** [comment]
On second thought, maybe we should just use #17420  itself as it already has some ongoing discussion.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/config/TaskLockConfig.java [hunk 1]
```diff
@@ -36,6 +36,9 @@ public class TaskLockConfig
   @JsonProperty
   private long batchAllocationWaitTime = 0L;
 
+  @JsonProperty
+  private boolean batchAllocationReduceMetadataIO = false;
```
Comment (cryptoe) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/config/TaskLockConfig.java:40]: Should this be always on and then we remove this config in next release ?

Comment (kfaraz) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/config/TaskLockConfig.java:40]: Yeah, I guess the changes are safe enough. We can always keep it on. Let me try to run ITs with it turned on.

**Review by cryptoe — APPROVED** [review]

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/config/TaskLockConfig.java (hunk 1 above)
Comment (kfaraz) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/config/TaskLockConfig.java]: Flag has been set to true by default.

**Review by AmatyaAvadhanula — APPROVED** [review]

**Comment by kfaraz** [comment]
Thanks for the reviews, @AmatyaAvadhanula , @cryptoe !
Taking a final look at the PR, then I will merge the changes.

File: server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java [hunk 2]
```diff
@@ -2900,6 +2926,67 @@ public int removeDataSourceMetadataOlderThan(long timestamp, @NotNull Set<String
...
+        = SqlSegmentsMetadataQuery.forHandle(handle, connector, dbTables, jsonMapper)
+                                  .retrieveUsedSegmentIds(dataSource, interval);
+    // Map from version -> interval -> segmentId with the smallest partitionNum
+    Map<String, Map<Interval, SegmentId>> versionIntervalToSmallestSegmentId = new HashMap<>();
+    for (SegmentId segmentId : overlappingSegmentIds) {
+      final Map<Interval, SegmentId> map
+          = versionIntervalToSmallestSegmentId.computeIfAbsent(segmentId.getVersion(), v -> new HashMap<>());
+      final SegmentId value = map.get(segmentId.getInterval());
+      if (value == null || value.getPartitionNum() > segmentId.getPartitionNum()) {
+        map.put(interval, segmentId);
```
Comment (gianm) [inline server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java:2946]: I think the key here should be `segmentId.getInterval()` not `interval`. @kfaraz does that sound right to you?

Comment (kfaraz) [reply server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java:2946]: Yes, thanks a lot for catching this, @gianm !

**Pull request merged**